apis/_codegen/fetch-apis.go

```diff
@@ -1,99 +1,235 @@
+	source, err := gen.Format()
 	if err != nil {
-		fmt.Println("go fmt, code generation failed, error: ", err)
-		os.Exit(1)
+		log.Println(gen.String())
+		log.Fatalln("error: go fmt, code generation failed: ", err)
 	}
 
 	if err := ioutil.WriteFile("services.go", source, 0664); err != nil {
-		fmt.Println("Failed to save services.go, error: ", err)
-		os.Exit(1)
+		log.Fatalln("error: failed to save services.go: ", err)
 	}
 }
+
+func fetchService(g *got.Got, name string, url string) definitions.Service {
+	fmt.Println(" - Fetching", name)
+	// Fetch reference
+	res, err := g.Get(url).Send()
+	if err != nil {
+		log.Fatalln("error: failed to fetch API ", name, ": ", err)
+	}
+	// Parse reference
+	var s definitions.Service
+	if err := json.Unmarshal(res.Body, &s); err != nil {
+		log.Fatalln("error: failed parse API ", name, ": ", err)
+	}
+	return s
+}
+
+func fetchSchema(g *got.Got, url string) string {
+	fmt.Println(" - ", url)
+	res, err := g.Get(url).Send()
+	if err != nil {
+		log.Fatalln("error: failed to fetch ", url, ": ", err)
+	}
+	// Test that we can parse the JSON schema (otherwise it's invalid)
+	var i interface{}
+	if err := json.Unmarshal(res.Body, &i); err != nil {
+		log.Fatalln("error: failed to parse ", url, ": ", err)
+	}
+	return string(res.Body)
+}
+
+// generator holds a buffer of the output that will be generated.
+type generator struct {
+	buf bytes.Buffer
+}
+
+func (g *generator) Write(p []byte) (n int, err error) {
+	return g.buf.Write(p)
+}
+
+// Printf prints the given format+args to the buffer.
+func (g *generator) Printf(format string, args ...interface{}) {
+	fmt.Fprintf(&g.buf, format, args...)
+}
+
+// Print prints the given a to the buffer.
+func (g *generator) Print(a ...interface{}) {
+	fmt.Fprint(&g.buf, a...)
+}
+
+func (g *generator) PrintComposite(data interface{}, sorted bool) {
+	v := reflect.ValueOf(data)
+	t := v.Type()
+
+	switch v.Kind() {
+	case reflect.Array, reflect.Slice:
+		// do stuff
+		g.Printf("%s", t.String())
+		if v.Kind() == reflect.Slice && v.IsNil() {
+			g.Printf("(nil)")
+			break
+		}
+		g.Print("{\n")
+		if sorted && t.Elem().Kind() == reflect.String {
+			sortedK := make([]string, 0, v.Len())
+			for i := 0; i < v.Len(); i++ {
+				sortedK = append(sortedK, v.Index(i).String())
+			}
+			sort.Strings(sortedK)
+			for _, s := range sortedK {
+				g.Printf("%#v,\n", s)
+			}
+		} else {
+			for i := 0; i < v.Len(); i++ {
+				g.PrintComposite(v.Index(i).Interface(), sorted)
+				g.Print(",\n")
+			}
+		}
+		g.Print("}")
+	case reflect.Struct:
+		// then struct
+		g.Printf("%s{\n", t.String())
+		for i := 0; i < v.NumField(); i++ {
+			g.Printf("%s: ", t.Field(i).Name)
+			g.PrintComposite(v.Field(i).Interface(), sorted)
+			g.Print(",\n")
+		}
+		g.Print("}")
+	case reflect.Map:
+		// finally map
+		g.Printf("%s{\n", t.String())
+		keys := v.MapKeys()
+		if len(keys) == 0 {
+			g.Print("}")
+			break
+		}
+		if sorted && t.Key().Kind() == reflect.String {
```

Comment:
I would argue that sorting always makes sense here, regardless of `sorted`.  The nondeterminism we want to avoid is in map ordering, rather than in array order in upstream schema/reference generation.